src/lib.rs

```diff
@@ -0,0 +1,1036 @@
+        Self { bits }
+    }
+    #[inline(always)]
+    pub fn before(inst: Inst) -> Self {
+        Self::new(inst, InstPosition::Before)
+    }
+    #[inline(always)]
+    pub fn after(inst: Inst) -> Self {
+        Self::new(inst, InstPosition::After)
+    }
+    #[inline(always)]
+    pub fn inst(self) -> Inst {
+        // Cast to i32 to do an arithmetic right-shift, which will
+        // preserve an `Inst::invalid()` (which is -1, or all-ones).
+        Inst::new(((self.bits as i32) >> 1) as usize)
+    }
+    #[inline(always)]
+    pub fn pos(self) -> InstPosition {
+        match self.bits & 1 {
+            0 => InstPosition::Before,
+            1 => InstPosition::After,
+            _ => unreachable!(),
+        }
+    }
+    #[inline(always)]
+    pub fn next(self) -> ProgPoint {
+        Self {
+            bits: self.bits + 1,
+        }
+    }
+    #[inline(always)]
+    pub fn prev(self) -> ProgPoint {
+        Self {
+            bits: self.bits - 1,
+        }
+    }
+    #[inline(always)]
+    pub fn to_index(self) -> u32 {
+        self.bits
+    }
+    #[inline(always)]
+    pub fn from_index(index: u32) -> Self {
+        Self { bits: index }
+    }
+}
+
+/// An instruction to insert into the program to perform some data movement.
+#[derive(Clone, Debug)]
+pub enum Edit {
+    /// Move one allocation to another. Each allocation may be a
+    /// register or a stack slot (spillslot). However, stack-to-stack
+    /// moves will never be generated.
+    ///
+    /// `to_vreg`, if defined, is useful as metadata: it indicates
+    /// that the moved value is a def of a new vreg.
+    ///
+    /// `Move` edits will be generated even if src and dst allocation
+    /// are the same if the vreg changes; this allows proper metadata
+    /// tracking even when moves are elided.
+    Move {
+        from: Allocation,
+        to: Allocation,
+        to_vreg: Option<VReg>,
+    },
+    /// Define blockparams' locations. Note that this is not typically
+    /// turned into machine code, but can be useful metadata (e.g. for
+    /// the checker).
+    BlockParams {
+        vregs: Vec<VReg>,
+        allocs: Vec<Allocation>,
+    },
+    /// Define a particular Allocation to contain a particular VReg. Useful
+    /// for the checker.
+    DefAlloc { alloc: Allocation, vreg: VReg },
+}
+
+/// A machine envrionment tells the register allocator which registers
+/// are available to allocate and what register may be used as a
+/// scratch register for each class, and some other miscellaneous info
+/// as well.
+#[derive(Clone, Debug)]
+pub struct MachineEnv {
+    /// Physical registers. Every register that might be mentioned in
+    /// any constraint must be listed here, even if it is not
+    /// allocatable under normal conditions.
+    pub regs: Vec<PReg>,
+    /// Preferred physical registers for each class. These are the
+    /// registers that will be allocated first, if free.
+    pub preferred_regs_by_class: Vec<Vec<PReg>>,
```

Comment:
Since the number of registers classes is hard-coded at 2, this should just be `[Vec<PReg>; 2]`.